### PR TITLE
fix not updating size of array

### PR DESCRIPTION
**Pull request by paul-szczepanek-arm**
the other change is to respect existing elements
code reordered for clarity

### CODE REVIEW

File: features/FEATURE_BLE/ble/generic/SecurityDb.h [hunk 1]
```diff
@@ -584,28 +584,32 @@ class SecurityDb {
...
                 continue;
             }
 
+            SecurityEntryIdentity_t* identity = read_in_entry_peer_identity(db_handle);
+            if (!identity) {
+                continue;
+            }
+
+            memcpy(
+                whitelist->addresses[whitelist->size].address,
```
Comment (paul-szczepanek-arm) [inline features/FEATURE_BLE/ble/generic/SecurityDb.h]: indexed by size to respect existing elements

File: features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp [hunk 2]
```diff
@@ -142,6 +142,9 @@ ble_error_t GenericSecurityManager::purgeAllBondingState(void) {
 ble_error_t GenericSecurityManager::generateWhitelistFromBondTable(Gap::Whitelist_t *whitelist) const {
     if (!_db) return BLE_ERROR_INITIALIZATION_INCOMPLETE;
     if (eventHandler) {
+        if (!whitelist) {
```
Comment (paul-szczepanek-arm) [inline features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp]: now wee need to check it's not NULL

File: features/FEATURE_BLE/ble/generic/SecurityDb.h [hunk 3]
```diff
@@ -584,28 +584,32 @@ class SecurityDb {
         WhitelistDbCb_t cb,
         ::Gap::Whitelist_t *whitelist
     ) {
-        for (size_t i = 0; i < get_entry_count() && i < whitelist->capacity; i++) {
+        for (size_t i = 0; i < get_entry_count() && whitelist->size < whitelist->capacity; i++) {
```
Comment (donatieng) [inline features/FEATURE_BLE/ble/generic/SecurityDb.h]: Should `whitelist->size` be initialised to 0? Or is it the user's responsibility?

File: features/FEATURE_BLE/ble/generic/SecurityDb.h [hunk 4]
```diff
@@ -584,28 +584,32 @@ class SecurityDb {
...
             entry_handle_t db_handle = get_entry_handle_by_index(i);
             SecurityDistributionFlags_t* flags = get_distribution_flags(db_handle);
 
             if (!flags) {
                 continue;
             }
 
+            SecurityEntryIdentity_t* identity = read_in_entry_peer_identity(db_handle);
+            if (!identity) {
```
Comment (donatieng) [inline features/FEATURE_BLE/ble/generic/SecurityDb.h]: Can that be checked against `NULL`?

File: features/FEATURE_BLE/ble/generic/SecurityDb.h (hunk 3 above)
Comment (paul-szczepanek-arm) [reply features/FEATURE_BLE/ble/generic/SecurityDb.h]: whitelist is user provided, could already have something in it, would be 0 on a newly created one

File: features/FEATURE_BLE/ble/generic/SecurityDb.h (hunk 4 above)
Comment (paul-szczepanek-arm) [reply features/FEATURE_BLE/ble/generic/SecurityDb.h]: the check is completely equivalent and down to style, both styles are used in mbed

Comment (donatieng) [reply features/FEATURE_BLE/ble/generic/SecurityDb.h]: I don't like this style :). But ok ;)

**Pull request merged**